runtime/src/iree/modules/hal/module.c

```diff
@@ -945,6 +945,212 @@ IREE_VM_ABI_EXPORT(iree_hal_module_command_buffer_dispatch_indirect,  //
       command_buffer, executable, entry_point, workgroups_ref, flags);
 }
 
+// Argument signature: rriiiiICiDCiirIID
+typedef struct {
+  union {
+    struct {
+      iree_vm_ref_t command_buffer;
+      iree_vm_ref_t executable;
+      int32_t entry_point;
+      uint32_t workgroup_count[3];
+      iree_hal_dispatch_flags_t flags;
+    };
+    iree_vm_abi_rriiiiI_t params;
+  };
+  iree_vm_size_t constant_count;
+  const uint32_t* constants;
+  iree_vm_size_t binding_count;
+  const iree_vm_abi_iirII_t* bindings;
+} iree_hal_module_command_buffer_dispatch2_args_t;
+static iree_status_t iree_hal_module_command_buffer_dispatch2(
```

Comment:
Oof this is intimidating / tricky to read. Not sure how to improve it though.

Reply:
yeah, prior to this the only other place doing it was in the inline HAL (for this exact reason) - hopefully it discourages others from designing APIs that require multiple variadics a bit :P